File: C/binomial_coefficient.c
// binomial_coefficient.c
// Implementation of the binomial coefficient calculation using Pascal's Triangle

#include "binomial_coefficient.h"
#include <stdio.h>

int binomial_coefficient(int n, int k) {
    if (k < 0 || k > n) return 0;

    int pascal_triangle[n + 1][n + 1];

    // Construct Pascal's Triangle
    for (int i = 0; i <= n; ++i) {
        pascal_triangle[i][0] = 1; // C(i, 0) = 1
        for (int j = 1; j <= i; ++j) {
            pascal_triangle[i][j] = pascal_triangle[i - 1][j - 1] + pascal_triangle[i - 1][j];
        }
    }

    return pascal_triangle[n][k];
}

File: Ruby/binomial_coefficient.rb
# binomial_coefficient.rb
# Function to calculate the binomial coefficient C(n, k) using Pascal's Triangle

class BinomialCoefficient
  def self.calculate(n, k)
    return 0 if k < 0 || k > n

    # Initialize Pascal's Triangle
    pascal_triangle = Array.new(n + 1) { Array.new(n + 1, 0) }

    # Build Pascal's Triangle
    (0..n).each do |i|
      pascal_triangle[i][0] = 1 # C(i, 0) = 1
      (1..i).each do |j|
        pascal_triangle[i][j] = pascal_triangle[i - 1][j - 1] + pascal_triangle[i - 1][j]
      end
    end

    # Return the binomial coefficient
    pascal_triangle[n][k]
  end
end
